### ISSUE TITLE
Gradient of tf.pow

### ISSUE THREAD

**Issue by julianoks** [tensorflow/tfjs#346]
There are two issues that I'd like to point out with tf.pow's gradient function:

First, there's bug with the gradient of tf.pow w.r.t. the exponent, whereby a non-positive base will result in a NaN for the corresponding exponent's gradient. This can be reproduced using: 
```
tf.version_core == "0.8.1";
tf.grad(x => tf.pow(tf.scalar(0), x))(tf.scalar(2)).dataSync();
tf.grad(x => tf.pow(tf.scalar(-2), x))(tf.scalar(2)).dataSync();
```
The backend pow is [implemented as](https://github.com/tensorflow/tfjs-core/blob/master/src/kernels/webgl/binaryop_gpu.ts#L50) `pow(abs(a), b) * isEven(round(a))`, where `isEven` is an indicator function, `isEven: X \to {1,-1}`. Despite the backend, the [gradient function](https://github.com/tensorflow/tfjs-core/blob/master/src/ops/binary_ops.ts#L215) takes the log of the base, which may be negative, rather than the non-negative base used in the backend, which was hit with an abs. One solution is to have an "unprotected" backend pow (which only receives non-negative bases) and have the abs and the isEven components in the tf.pow op definition, which would be cached for the backward pass. Although this takes care of the gradient for when the base is negative, we still have an issue for when the base is 0.

When the base is 0, both the gradient of the base and the exponent will be NaN: 
```
tf.version_core == "0.8.1";
tf.grad(x => tf.pow(x, tf.scalar(2)))(tf.scalar(0)).dataSync(); // grad of base
tf.grad(x => tf.pow(tf.scalar(0), x))(tf.scalar(2)).dataSync(); // grad of exponent
```
As just explained, when the base is 0, the gradient will be NaN for the exponent because we're taking `log(0)`, and will be NaN for the base because we're [dividing by 0](https://github.com/tensorflow/tfjs-core/blob/master/src/ops/binary_ops.ts#L206). A simple solution would be to zero-out the gradient of both the base and the exponent wherever the base is 0.
This makes sense for the gradients (for a^x when a=0) because
* the derivative w.r.t. the exponent will be a^x * log(a) = 0^x*log(0) = 0 * log(0)
* the derivative w.r.t. the base will be x*0^(x-1) = 0

Fortunately, I believe that using the exponent's gradient is a rare use case because there's rarely a path from a variable to an exponent (am I wrong to assume this?), so fixing the gradient of the exponent wouldn't be such a high priority. However, fixing the gradient of the base should be done because it is commonly encountered, eg in regularization; if someone is using tf.pow (and not tf.square) and a parameter reaches 0, then a NaN will be produced. This will eventually propagate to variables and mess everything up.

Would y'all be open to these changes?

**Comment by jgartman** [comment]
@julianoks I think you're right that there's a problem in the case that the base is 0 and the gradient is being taken with respect to the exponent.  Like you said, It will be `nan` in all cases as currently computed.  However, it should only be `nan` if the exponent is not positive.  I recommend the calculation of that gradient be changed to:
```
dy.mul(exp.toFloat().mul(y.pow(base, exp - 1)));
```

I very easily could be missing the problem but I don't think the gradient is wrong in the first case you mentioned though.  As was pointed out, the backend is computing `pow` as:
```
return (round(mod(b, 2.0)) == 0 || round(mod(b, 2.0)) == 2) ?
      pow(abs(a), b) : sign(a) * pow(abs(a), b)
```
As I understand it, the reason for taking the absolute value of the base and checking the parity of the exponent is just because the behavior of the GLSL `pow` function is undefined when the base is negative. So, for example suppose we're doing `pow(a, b)` where `a` is negative and `b` is odd.  Instead of doing `pow(a, b)` where the result would be undefined it's actually being computed as `-1 * pow(abs(a), b)`.  The function being computed is still `a^b` in all cases and the gradient with respect to the base is still `a^b ln a`.  Since non-positive values of `a` are outside the domain of the function`a^b ln a` the gradient with respect to `a` should be `nan` where `a` is negative or 0.

**Comment by julianoks** [comment]
@jgartman 
The forward pass only throws NaNs when the base is 0 and the exponent is non-positive, so having the gradient reflect that by throwing a NaN is reasonable. Although I agree with your suggestion to change the base's gradient from `b*(a^b/a)` to `b*(a^(b-1))`, to handle the case when `b=0`, I find zeroing out the gradient at `b=0` appealing because it removes NaNs in a (seemingly) innocuous way (yet if the forward pass produces NaNs, it may poison upstream gradients).

Like you mentioned, tfjs uses a "protected" version of the pow function for negative bases, effectively extending the domain of the forward pass to include negative bases, such as `(-5)^2`. What I was pointing out was that the domains of the forward and backwards pass are inconsistent; specifically that the exponent's gradient is NaN when the base is negative:
```
(x => tf.pow(tf.scalar(-5),x))(tf.scalar(2)).dataSync(); // forward pass is OK
tf.grad(x => tf.pow(tf.scalar(-5),x))(tf.scalar(2)).dataSync(); // but backwards pass returns NaN
```

I disagree with your final statement where you said:

> The function being computed is still `a^b` in all cases and the gradient with respect to the exponent is still `a^b ln a`. Since non-positive values of a are outside the domain of the function `a^b ln a` the gradient with respect to a should be nan where a is negative or 0.

The function being computed in the forward pass is a protected version of pow, which is not `a^b` in all cases. I'm suggesting that the gradient function be aligned with the protected pow function used in the forward pass.

**Comment by jgartman** [comment]
@julianoks  First I should clarify and correct a couple of things from my previous post.  I think you understood this but just to be clear, when I was using the notation `a^b` I was referring to the mathematical power function and when I was using `pow()` I was referring to the tfjs or GLSL implementations of that function.  My statement that the tfjs implementation matches the mathematical power function in all cases is wrong, specifically for negative bases and non integer exponents since the library doesn't support complex numbers.

With that out of the way, I went and checked what TF does for the `pow` gradient since maintaining consistency between the two libraries is usually a goal and it looks like they just return 0 for negative bases https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/math_grad.py#L945..  I think that sounds like a pretty reasonable thing for us to do here, again given the lack of support for complex numbers.

**Comment by julianoks** [comment]
@jgartman 
I agree that TF's python implementation of the pow gradient function would be good to follow. 
That implementation literally says `"There's no sensible real value to return if x < 0, so return 0"`, which I'm still somewhat wary of because it doesn't reflect the gradient of the implemented forward function, regardless of whether the forward function is deemed sensible or not.
Ultimately, I don't think the gradient of the exponent is of such practical importance, although I'd prefer it to return 0's instead NaNs when the base is non-positive (like in TF python).

**In conclusion, we should change the base's gradient to `b*(a^(b-1))` and zero-out the gradient of the exponent wherever the base is non-positive.** If @dsmilkov would be willing to accept these changes, I'd be happy to make a PR.

**Comment by jgartman** [comment]
Opened a separate issue for the forward func #350.

**Comment by jgartman** [comment]
@dsmilkov Just wanted to bump this issue since I noticed it was still open and assigned to you.  Were you planning on opening a PR for this or should Julian go ahead with his proposed fix?

**Comment by dsmilkov** [comment]
Thanks for bumping this up. If the proposed change will align TF.js to the behavior with TF's Python, then yes, Julian should go ahead with the proposed fix. I'm happy to review. Thanks!

**Comment by nsthorat** [comment]
This is fixed!
